028-service-binding.md

```diff
@@ -0,0 +1,340 @@
+
+## Binding to a Kafka cluster with no TLS or authentication
+
+The bootstrapServers address for clients to use is currently in the `status` of the `Kafka CR`:
+
+``` yaml
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: my-cluster-kafka-bootstrap.my-namespace.svc
+      port: 9092
+    bootstrapServers: my-cluster-kafka-bootstrap.my-namespace.svc:9092
+  - type: tls
+    addresses:
+    - host: my-cluster-kafka-bootstrap.my-namespace.svc
+      port: 9093
+    bootstrapServers: my-cluster-kafka-bootstrap.my-namespace.svc:9093
+```
+
+The problem with the location of this data is that it is not contained in a `Secret` and is therefore not accessible to a Service Binding Operator. Since there are also multiple listeners to choose from it is not obvious which should be placed in a `Secret` or how the application chooses the correct listener to call.
+
+## Binding to a Kafka cluster with TLS but no authentication
+
+The addition with this scenario is that Kafka clients need access to the CA certificate that signed the broker's server certificate.
+
+The CA certificate is most easily obtained from the `<cluster>-cluster-ca-cert` secret. While this name is predictable, it is not known to the Service Binding Operator. In addition the application also needs the bootstrapServer information, but that is not contained in the currently provided `Secret`.
+
+The consuming client needs to know the following binding information:
+
+* **bootstrapServer** - bootstrap server information for the listener it is contacting
+* **ca.p12** - CA certificate PKCS #12 archive file for storing certificates and keys
+* **ca.password** - password for protecting the CA certificate PKCS #12 archive file
+* **ca.crt** - CA certificate for the cluster
+
+This information is spread across the `Kafka` CR and the `Secret`.
+
+## Binding to a Kafka cluster with username/password authentication
+
+Strimzi provides the `KafkaUser` custom resource as a way of managing users and credentials. Using the SASL SCRAM mechanism, the consuming client's credentials are made available in a combination of the CR's status and a secret.
+
+The `KafkaUser` CR contains information about the user and the secret with the password. For example:
+
+``` yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaUser
+metadata:
+  name: my-user
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  authentication:
+    type: scram-sha-512
+  authorization:
+    type: simple
+    acls:
+    - resource:
+        type: topic
+        name: my-topic
+        patternType: literal
+      operation: Read
+status:
+  username: my-user-name
+  secret: my-user
+```
+
+The secret looks like this:
+
+``` yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-user
+  labels:
+    strimzi.io/kind: KafkaUser
+    strimzi.io/cluster: my-cluster
+type: Opaque
+data:
+  password: Z2VuZXJhdGVkcGFzc3dvcmQ=
```

Comment:
For the completeness, it has the JAAS string now as well.

Reply:
@ppatierno @scholzj How is this field designed to be used? Most commonly in Java applications I have seen the following configuration:
```
sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username="<username>" password="<password>";
```

Reply:
the content of the field is exactly something like you showed above. See the example in the official Strimzi doc [1].
In your Java application you can just get the value of this field and setting it for the "sasl.jaas.config" property (in your Properties map). From that example, if you decode the base64 value you get "org.apache.kafka.common.security.scram.ScramLoginModule required username="my-user" password="generatedpassword";

[1] https://strimzi.io/docs/operators/latest/using.html#scram_sha_512_authentication

Reply:
Ah yes, I forgot to base64 decode the value that I was looking at in my secret!